### ISSUE TITLE
Optimize simple data implementations to reduce memory footprint

### ISSUE THREAD

**Issue by hdbeukel** [corehunter/corehunter3#72]
It may be useful to provide low memory data implementations to deal with very large datasets. For example we could

- [x] Use primitive type matrices with reserved values for missing data (e.g. `-128` for allele score and `NaN` for distances and frequencies).
- [x] Not precompute and store the frequencies in `BiAllelicGenotypeData`, but recompute from allele scores on demand (may increase runtime).

Could be interesting to provide a memory/runtime tradeoff to the user.

**Comment by daveneti** [comment]
This was one of the motivations of 'extended' module. Have there been any reason for needing these implementations?

**Comment by hdbeukel** [comment]
I was contacted by a researcher from CIMMYT who is dealing with a 10.000 x 55.500 marker dataset and they were running into memory issues. I explained how to increase the amount of memory reserved for Java, i.e. with
```
options(java.parameters = "-Xmx...g")
```
which should work for them (they have 16GB RAM), so this is not an urgent issue. Still, it would make things easier and allow users having less memory available to analyze large datasets.

**Comment by hdbeukel** [comment]
I also suggested to precompute a distance matrix if they do not want to maximize allelic richness.

**Comment by hdbeukel** [comment]
Btw, I guess that especially precomputing the frequencies for bi-allelic data may be a waste of memory. The allele scores require only 4 bytes (integer) per data point, while the two inferred frequencies (one for each allele) together require 16 bytes (two doubles). Going from 4 to 20 bytes seems a lot, no? Perhaps it is better then to recompute frequencies from the allele scores and accept a slight increase in computation time for a large decrease in memory?

**Comment by daveneti** [comment]
Ok, thanks. Can you tell me who at CIMMYT? Send me a mail if you like.

**Comment by guypfr** [comment]
BTW was this using version 2 or 3?

**Comment by hdbeukel** [comment]
I got a lot of question when presenting our poster at PAG about the size of the datasets that can be analyzed. All datasets in the paper are small and we actually have no idea about the limits, do we? I suggest we make the `SimpleBiallelicGenotypeData` more memory efficient as explained above and  experiment with large SNP datasets (random if need be) to assess the memory limits.

I had people talking about datasets with multiple 100k accessions and markers.

**Comment by hdbeukel** [comment]
We could also use the `byte` type to store the 0/1/2 allele scores instead of `int`. That would decrease the memory with a factor of 4.

**Comment by daveneti** [comment]
You are right is was never tested for large datasets. The extended module (never used) was the place we considered for optimised data structures. I had plans for an HDF (or similar implementation). Time to resurrect these plans.

**Comment by hdbeukel** [comment]
Ok so advanced implementations like HDF could go in the extended module but shouldn't we optimize the simple implementation as much as possible as well? I don't see the point in using `Integer` objects when we can reduce the memory usage with a factor 10 by just using `byte` types.

**Comment by hdbeukel** [comment]
So what do we conclude now? I suggest:

1. Optimize the current implementations while sticking to the simple data structures (version 3.2).
2. Explore more advanced data structures in the extended module (later).

On the way we can experiment with large random SNP datasets of various sizes to see how far we can go and how much we gain from the optimizations or complex data structures. If you agree the above steps I will split this into two separate issues.

**Comment by hdbeukel** [comment]
To convince you that the current implementations are really wasting a lot of memory without any need, I did a quick estimation of the number of bytes used per data point in the biparental format. The allele score takes at least 12 bytes (8 for object housekeeping, 4 for integer value) and the inferred frequencies of the two alleles each require at least 16 bytes (8 for housekeeping, 8 for double value). That gives 2*16 + 12 = 44 or likely even more bytes per data point.

I think this is way too much even for the simple implementations and we could easily reduce the memory usage by a massive amount to one byte per data point, just by using primitive `byte` type values and not precomputing the frequencies.

I did some quick tests and the current implementation already runs into trouble on my laptop with 16GB for a dataset with 10k individuals and 10k markers, which isn't even considered large these days. So I am even more convinced now we should follow the two steps outlined in the previous comment. Agreed?

**Comment by daveneti** [comment]
Hi Herman, I did not need any convincing. These simple implementations were only short term solutions in any case.

**Comment by hdbeukel** [comment]
Ok great. I would like to have a look at the first step (optimizing the current implementations) and later we can discuss ideas for advanced data structures.

**Comment by hdbeukel** [comment]
I optimized the simple data structures. Seems to be a great improvement! I was able to load and analyze datasets with multiple thousands accessions and up to 100k markers.

For those large datasets, it does take quite some time to load the data. I guess rJava has a lot of work there to transfer all the data from R to Java, with the necessary conversions etc. But it works fine! I reserved 8GB RAM for Java, no memory exceptions at all.
